src/shared/tpm2-util.c

```diff
@@ -1664,18 +1901,39 @@ int tpm2_unseal(const char *device,
         if (r < 0)
                 return r;
 
+        /* If their is a primary key we trust, like an SRK, use it */
         _cleanup_tpm2_handle_ Tpm2Handle *primary = NULL;
-        r = tpm2_make_primary(c, &primary, primary_alg, NULL);
-        if (r < 0)
-                return r;
+        if (srk_buf) {
+
+                r = tpm2_handle_new(c, &primary);
+                if (r < 0)
+                        return r;
+
+                primary->flags |= TPM2_HANDLE_FLAGS_NOFLUSH;
+
+                log_debug("Found existing SRK key to use, deserializing ESYS_TR");
+                rc = sym_Esys_TR_Deserialize(
```

Comment:
sorry to jump in again, but could you please marshal this instead of serializing? If the 'srk' is formatted as a serialized ESYS handle, then it's impossible to do this with an 'external' TPM. And you don't really gain anything by serializing instead of marshalling, as you are restricting the srk to only live at the specific reserved srk handle address, so you really don't need to retain that detail.

Basically, if this is merged using serialization then it won't be useful at all to my patch series, and it will be yet another optional parameter that tpm2_unseal() will have to support. I would really much prefer my approach of saving the marshalled key/template that I have queued up for a PR: https://github.com/ddstreet/systemd/commits/tpm2_save_primary_template

Reply:
> sorry to jump in again, but could you please marshal this instead of serializing?

This is the marshalling and unmarshalling routines for an ESYS_TR handle

> If the 'srk' is formatted as a serialized ESYS handle, then it's impossible to do this with an 'external' TPM.

That's not true. ESAPI and tpm2-tools allow for conversions between raw `TPM2_HANDLE`'s (eg 0x81000001) and ESYS_TR representations.

> And you don't really gain anything by serializing instead of marshalling, as you are restricting the srk to only live at the specific reserved srk handle address, so you really don't need to retain that detail.

You're actually losing the whole point of serializing and deserializing an ESYS_TR. An ESYS_TR is both the `TPM2_HANDLE` and most importantly the **objects name, which prevents spoofing**. This is what provides the binding for the salted session where no one can spoof the object.

I don't see any point in saving the template, everything should move to a fixed persistent key at a handle which provide t he greatest flexibility in implementation. That gets you to a point where you don't even ever need to care about templates, any key without a password could be the persistent parent. You could shove in any ESYS_TR to use a non-standard location as well.

Reply:
@ddstreet sent you an invite to your ieee.org email for a meeting, figured easier to discuss verbally than these long threads.

Reply:
TL;DR
1. Creating an offline seal key TPM2B_PUBLIC and PRIVATE blobs still requires a stable parent key to apply duplication protections
    - The creating entity needs to know the expected parent on the actual TPM which is dependent on the Owner Hierarchy Seed. 
2. The primary parent key at some point needs to be resolved to an ESYS_TR, which can be done ahead of time as well, these are stable handles.
    - It's the TPM_HANDLE + name, since the parent needs to be stable, this can be generated ahead of time as needed.
    - They can be moved around (hence deserialize and serialize). However that exact object needs to be resident in wherever it's moved to.
3. Moving systemd to a model that prevents use of owner hierarchy directly and thus owner auth is ideal, as it supports workloads where owner auth and disk enrolling entity are different.